### PR TITLE
    Added plugin that allows to mask event by regular expression

### PR DESCRIPTION
**Pull request by mmatros**
and replace matched on template-substitution

### CODE REVIEW

File: plugin/action/mask/mask.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+	case node.IsArray():
+		for _, n := range node.AsArray() {
+			applyForStrings(n, applyFn)
+		}
+	case node.IsObject():
+		for _, n := range node.AsFields() {
+			applyForStrings(n, applyFn)
+		}
+	default:
+		if node.IsString() {
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:142]: here you can remove if, AsString converts any JSON type to string

File: plugin/action/mask/mask.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+func (p *Plugin) maskAll(value string) (*string, bool) {
+	for i := range p.config.Masks {
+		value = p.re[i].ReplaceAllString(value, p.config.Masks[i].Substitution)
+	}
+	return &value, true
+}
+
+func applyForStrings(node *insaneJSON.Node, applyFn func(string) (*string, bool)) {
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:129]: Let's build a flat list of nodes (maybe add this function to insane-json lib).
And name this function `getValueNodeList`
And when just iterate by list and make substitution

File: plugin/action/mask/mask.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	for i, mask := range p.config.Masks {
+		p.logger.Infof("compiling regexp %s/ - %s", mask.Re, mask.Substitution)
+		p.re[i] = regexp.MustCompile(mask.Re)
+	}
+}
+
+func (p *Plugin) Stop() {
+}
+
+func (p *Plugin) maskByIndex(value string) (*string, bool) {
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:92]: here you don't need a pointer to a string, also you should return []byte, as insane-json has fast function to deal with bytes

File: plugin/action/mask/mask.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+			isMasked = true
+			offset += len(value) - l
+			l = len(value)
+			p.logger.Infof("mask by index %s", mask.Substitution)
+		}
+	}
+	return &value, isMasked
+}
+
+func (p *Plugin) maskIfMatched(value string) (*string, bool) {
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:110]: is it only for tests?

File: plugin/action/mask/mask.go [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+	case currentNode.IsArray():
+		for _, n := range currentNode.AsArray() {
+			collectValueNodes(n, valueNodes)
+		}
+	case currentNode.IsObject():
+		for _, n := range currentNode.AsFields() {
+			collectValueNodes(n, valueNodes)
+		}
+	default:
+		*valueNodes = append(*valueNodes, currentNode)
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:176]: Just pass valueNodes slice without pointer and return new slice (just like standard append())

File: plugin/action/mask/mask.go [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+	}
+}
+
+func getValueNodeList(rootNode *insaneJSON.Node) []*insaneJSON.Node {
+	var nodes []*insaneJSON.Node
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:181]: you can reuse node slice, put it into plugin

File: plugin/action/mask/mask.go [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+	collectValueNodes(rootNode, &nodes)
+	return nodes
+}
+
+func (p Plugin) Do(event *pipeline.Event) pipeline.ActionResult {
+	node := event.Root.Node
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:187]: node => root

File: plugin/action/mask/mask.go [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+	if node == nil {
+		return pipeline.ActionPass
+	}
+
+	nodes := getValueNodeList(node)
+
+	for _, v := range nodes {
+		data := v.AsBytes()
+		isMasked := p.mask(data)
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:196]: return buff explitically, mask fn don't need dependency on plugin struct

File: plugin/action/mask/mask.go [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+* First call regex.Match, if matching successful then call regex.Replace (maskIfMatched)
+* First call regex.FindMatchingIndex, then replace finded symbol by index(maskByIndex)
+
+}*/
+
+type Plugin struct {
+	config *Config
+	re     []*regexp.Regexp
+	logger *zap.SugaredLogger
+	buff   []byte
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:42]: we are using `buf` name in this project

File: plugin/action/mask/mask.go [hunk 10]
```diff
@@ -57,14 +59,15 @@ type Mask struct {
 	//> @3@4@5@6
 	//>
 	//> Regular expression used for masking
-	Re string `json:"re2" default:"" required:"true"`
+	ReStr string `json:"re2" default:"" required:"true"`
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:62]: Another plugins use `Re` / `Re_` naming. `Re` is a string, `Re_` is a compiled regexp. And why it's `re2` in json?

File: plugin/action/mask/mask.go [hunk 11]
```diff
@@ -57,14 +59,15 @@ type Mask struct {
...
 
 	//> @3@4@5@6
 	//>
-	//>	Substitution mask
-	Substitution byte `json:"substitution" default:"*"`
-
+	//>
+	// Substitution byte `json:"substitution" default:"*"`
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:67]: Write good description and also fix annotations

File: plugin/action/mask/mask.go [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+func maskValue(value, buf []byte, re *regexp.Regexp, grp []int) []byte {
+	buf = buf[:0]
+	indexes := re.FindAllSubmatchIndex(value, -1)
+	if len(indexes) == 0 {
+		return buf
+	}
+
+	offset := 0
+	for _, index := range indexes {
+		for _, group := range grp {
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:126]: function parameter should be more descriptive than local variable

File: plugin/action/mask/mask.go [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+			if group*2+1 >= len(index) {
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:127]: why this formula?

File: plugin/action/mask/mask.go [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+		}
+	default:
+		valueNodes = append(valueNodes, currentNode)
+	}
+	return valueNodes
+}
+
+func (p *Plugin) Do(event *pipeline.Event) pipeline.ActionResult {
+	root := event.Root.Node
+	if root == nil {
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:162]: remove this if

File: plugin/action/mask/mask.go [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+	return &Plugin{}, &Config{}
+}
+
+func (p *Plugin) Start(config pipeline.AnyConfig, params *pipeline.ActionPluginParams) {
+	p.config = config.(*Config)
+	p.buf = make([]byte, 0, params.PipelineSettings.AvgLogSize)
+	p.valueNodes = make([]*insaneJSON.Node, 0)
+	p.logger = params.Logger
+	for i := range p.config.Masks {
+		p.logger.Infof("compiling re: %s, groups: %v", p.config.Masks[i].Re, p.config.Masks[i].Groups)
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:81]: we use style "compiling, re=%s, groups=%v"

File: plugin/action/mask/mask.go [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+func (p Plugin) Stop() {
+}
+
+func appendMask(src, dst []byte, begin, end int) (int, []byte) {
+	runeCounter := utf8.RuneCount(src[begin:end])
+	for j := 0; j < runeCounter; j++ {
+		dst = append(dst, substitution)
+	}
+	offset := len(src[begin:end]) - runeCounter
+	return offset, dst
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:95]: let's here invert order of params

File: plugin/action/mask/mask.go [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+func maskSection(src, dst []byte, begin, end int) (int, []byte) {
+	dst = append(dst, src[0:begin]...)
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:99]: You may use `src[:begin]`

File: plugin/action/mask/mask.go [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+	if len(indexes) == 0 {
+		return buf
+	}
+
+	offset := 0
+	for _, index := range indexes {
+		for _, grp := range groups {
+			maxIndexValue := grp*2 + 1
+			if maxIndexValue < 0 || maxIndexValue >= len(index) || index[maxIndexValue] == -1 {
+				l.Warnf("For re %s group %v doesn't exists", grp, re.String())
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:122]: don't use the capital first letter
this warn may flood logs, are you sure to log warn here?
also, pass buf as first param

File: plugin/action/mask/mask.go [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+	p.valueNodes = p.valueNodes[:0]
+	p.valueNodes = getValueNodeList(root, p.valueNodes)
+
+	for _, v := range p.valueNodes {
+		value := v.AsBytes()
+		isMasked := false
+		for _, mask := range p.config.Masks {
+			res := maskValue(value, p.buf, mask.Re_, mask.Groups, p.logger)
+			if len(res) > 0 {
+				p.logger.Infof("value masked by re %s, groups %v", mask.Re, mask.Groups)
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:167]: it will flood logs

File: plugin/action/mask/mask.go [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+	root := event.Root.Node
+
+	p.valueNodes = p.valueNodes[:0]
+	p.valueNodes = getValueNodeList(root, p.valueNodes)
+
+	for _, v := range p.valueNodes {
+		value := v.AsBytes()
+		isMasked := false
+		for _, mask := range p.config.Masks {
+			res := maskValue(value, p.buf, mask.Re_, mask.Groups, p.logger)
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:165]: here you should use buf instead `res`, or `buf` won't grow

File: plugin/action/mask/mask.go [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+			if len(res) > 0 {
+				p.logger.Infof("value masked by re %s, groups %v", mask.Re, mask.Groups)
+				isMasked = true
+				value = res
+			}
+		}
+		if isMasked {
+			v.MutateToBytes(value)
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:173]: here you should copy `buf`, because you reuse it

File: plugin/action/mask/mask.go [hunk 22]
```diff
@@ -0,0 +1,175 @@
...
+		return value
+	}
+
+	buf = buf[:0]
+
+	offset := 0
+	for _, index := range indexes {
+		for _, grp := range groups {
+			maxIndexValue := grp*2 + 1
+			if maxIndexValue < 0 || maxIndexValue >= len(index) || index[maxIndexValue] == -1 {
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:124]: how can `maxIndexValue` be less than 0?

File: plugin/action/mask/mask_test.go [hunk 23]
```diff
@@ -0,0 +1,350 @@
...
+		},
+		{
+			name:     "ID&text&card",
+			input:    []string{`{"field1":"Иванов Иван Иванович c картой 4445-2222-3333-4444 встал не с той ноги"}`},
+			expected: []string{`{"field1":"******************** c картой ****-****-****-**** встал не с той ноги"}`},
+			comment:  "only ID & card number masked",
+		},
+		{
+			name:     "ID&text&card",
+			input:    []string{`{"field1":"Иванов Иван Иванович c картами 4445-2222-3333-4444 и 4445-2222-3333-4444"}`},
```
Comment (vitkovskii) [inline plugin/action/mask/mask_test.go:245]: let's use English and write only a few tests for UTF-8

File: plugin/action/mask/mask.go [hunk 24]
```diff
@@ -0,0 +1,175 @@
...
+**Example:**
+```yaml
+pipelines:
+  example_pipeline:
+    ...
+    actions:
+    - type: mask
+      masks:
+        - mask:
+			re: "\b(\d{1,4})\D?(\d{1,4})\D?(\d{1,4})\D?(\d{1,4})\b"
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:26]: fix tabs

File: plugin/action/mask/mask.go (hunk 22 above)
Comment (mmatros) [reply plugin/action/mask/mask.go:124]: incorrect group number in pipeline configuration file

File: plugin/action/mask/mask.go [hunk 25]
```diff
@@ -0,0 +1,222 @@
...
+			return false
+		}
+		uniqueGrp[g] = exists
+	}
+	return true
+}
+
+func verifyGroupNumbers(groups []int, totalGroups int, logger *zap.SugaredLogger) []int {
+	if len(groups) == 0 {
+		logger.Panicf("groups is empty")
```
Comment (vitkovskii) [inline plugin/action/mask/mask.go:109]: Use `Fatal` as it's config error, not internal corruption

**Pull request merged**